feat: Introduce sync-option SkipDryRunOnMissingResource=true (#2873)

a-hat opened the pull request:
* Annotation `argocd.argoproj.io/sync-options: SkipDryRunOnMissingResource=true` can be set on any resource
* When set, then the dry run for this resource will be skipped when syncing, but only if the Group and Kind are currently not present in the cluster.
* This is analogous to the current handling of missing CRDs, but when the CRD is part of the same sync.
* Supports the use case where a CRD is created by a kubernetes controller inside the cluster in response to another CRD in the same sync.
* Closes #2873

Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [x] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [x] Optional. My organization is added to USERS.md.
* [x] I've signed the CLA and my build is green ([troubleshooting builds](https://argoproj.github.io/argo-cd/developer-guide/ci/)).

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/argoproj/argo-cd?pullRequest=3247) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/argoproj/argo-cd?pullRequest=3247) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/argoproj/argo-cd?pullRequest=3247) it.</sub>

a-hat commented:
Sorry, I used the wrong e-mail address

a-hat commented:
Hey @alexmt! What do you think about this change? This is a feature on our customers wish list. Thanks a lot for your time!

alexmt commented:
Hello @a-hat, thank you for your contribution!

Sorry for the delay with the review. The use case makes sense to me and sync-option annotation seems like the right choice.  We are rushing to finally publish 1.5.0 release. Please give me a couple more details to complete it and get back to this PR.

a-hat commented:
@alexmt @jannfis Great! Thanks for your feedback.
I updated the version in the documentation and rebased the branch. Let me know if I can do anything else.
